middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -27,29 +29,93 @@ type cont =
         handler_body : Flambda.Expr.t
       }
 
-type extra_info =
-  | Untag of Cmm.expression
-  | Boxed_number
+type extra_info = Untag of Cmm.expression
+
+(* Since to_cmm_primitive.ml depends on this file, and in this file, we need to
+   translate delayed/split primitives, we need to have access to the translation
+   primitive from to_cmm_primitive.ml, and we'll get them through this
+   record. *)
+
+type prim_res = extra_info option * R.t * Cmm.expression
+
+type ('env, 'prim, 'arity) prim_helper =
+  'env -> R.t -> Debuginfo.t -> 'prim -> 'arity
+
+type 'env trans_prim =
+  { nullary : ('env, P.nullary_primitive, prim_res) prim_helper;
+    unary : ('env, P.unary_primitive, Cmm.expression -> prim_res) prim_helper;
+    binary :
+      ( 'env,
+        P.binary_primitive,
+        Cmm.expression -> Cmm.expression -> prim_res )
+      prim_helper;
+    ternary :
+      ( 'env,
+        P.ternary_primitive,
+        Cmm.expression -> Cmm.expression -> Cmm.expression -> prim_res )
+      prim_helper;
+    variadic :
+      ('env, P.variadic_primitive, Cmm.expression list -> prim_res) prim_helper
+  }
 
 (* Delayed let-bindings (see the .mli) *)
 
-type binding =
+(* the binding kinds *)
+type simple = Simple
+
+type complex = Complex
+
+type _ inline =
+  | Do_not_inline : simple inline
+  | May_inline_once : simple inline
+  | Must_inline_once : complex inline
+  | Must_inline_and_duplicate : complex inline
+
+(* Note on the effects of splittable bindings:
+
+   The arguments are stored with their effects. This means that if we need to
+   split the binding, we can re-bind each argument with its correct effects.
+
+   The [prim_effects] field stores the effects of the primitive itself (the part
```

Comment:
This part of the comment at least is outdated.

Reply:
the outdated part has been removed